src/persist-client/src/internal/apply.rs

```diff
@@ -156,164 +157,212 @@ where
     >(
         &mut self,
         cmd: &CmdMetrics,
-        work_fn: WorkFn,
+        mut work_fn: WorkFn,
     ) -> Result<(SeqNo, Result<R, E>, RoutineMaintenance), Indeterminate> {
-        let ret = Self::apply_unbatched_cmd_locked(
-            &mut self.state,
-            cmd,
-            work_fn,
-            &self.cfg,
-            &self.metrics,
-            &self.shard_metrics,
-            &self.state_versions,
-        )
-        .await;
-        // TODO: Update any (upcoming) cached copies of state.
-        ret
+        loop {
+            let ret = Self::apply_unbatched_cmd_locked(
+                &self.state,
+                cmd,
+                &mut work_fn,
+                &self.cfg,
+                &self.metrics,
+                &self.shard_metrics,
+                &self.state_versions,
+            )
+            .await;
+
+            match ret {
+                Ok(Ok((seqno, Ok((res, new_state)), maintenance))) => {
+                    self.update_state(new_state);
+                    return Ok((seqno, Ok(res), maintenance));
+                }
+                Ok(Ok((seqno, Err(err), maintenance))) => {
+                    return Ok((seqno, Err(err), maintenance));
+                }
+                Ok(Err(err)) => {
+                    return Err(err);
+                }
+                Err(ExpectationMismatch(seqno)) => {
+                    cmd.cas_mismatch.inc();
+                    self.fetch_and_update_state(Some(seqno)).await;
+                }
+            }
+        }
     }
 
     async fn apply_unbatched_cmd_locked<
         R,
         E,
         WorkFn: FnMut(SeqNo, &PersistConfig, &mut StateCollections<T>) -> ControlFlow<E, R>,
     >(
-        state: &mut TypedState<K, V, T, D>,
+        state: &Arc<RwLock<TypedState<K, V, T, D>>>,
         cmd: &CmdMetrics,
-        mut work_fn: WorkFn,
+        work_fn: &mut WorkFn,
         cfg: &PersistConfig,
         metrics: &Metrics,
         shard_metrics: &ShardMetrics,
         state_versions: &StateVersions,
-    ) -> Result<(SeqNo, Result<R, E>, RoutineMaintenance), Indeterminate> {
+    ) -> Result<
+        Result<
+            (
+                SeqNo,
+                Result<(R, TypedState<K, V, T, D>), E>,
+                RoutineMaintenance,
+            ),
+            Indeterminate,
+        >,
+        ExpectationMismatch,
+    > {
```

Comment:
This feels like it might over the helper-struct/enum size threshold, if we don't have any other ideas for simplifying the type. (Maybe just one big enum? All gets handled in the same place anyways.)